File: atmel/avr/examples/09-3-uart-spi-nrf24l01+/main.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import sys
import os.path
sys.path.insert(1,os.path.normpath(sys.path[0]+"../../../../../../python"))

import premain
from utils import s2hex, hexdump
import xserial
import time


#  Description of nRF24L01+ registers
#    ('NAME', command/address code, number of bytes to read)
#
nrf_registers = [ ("CONFIG",	 '\x00', 1),
		  ("EN_AA",	 '\x01', 1),
		  ("EN_RX_ADDR", '\x02', 1),
		  ("SETUP_AW",	 '\x03', 1),
		  ("SETUP_RETR", '\x04', 1),
		  ("RF_CH",	 '\x05', 1),
		  ("RF_SETUP",	 '\x06', 1),
		  ("STATUS",	 '\x07', 1),
		  ("RX_ADDR_P0", '\x0A', 5),
		  ("RX_ADDR_P1", '\x0B', 5),
		  ("DYNPD",	 '\x1C', 1),
		  ("FEATURE",	 '\x1D', 1) ]


#  Parse command line, open the communication
#
import argparse
parser = argparse.ArgumentParser()
xserial.add_arguments(parser)
options = parser.parse_args()
if options.wires == 0:
    options.wires = 2
serial = xserial.get_serial(options) # resets the device


prompt = serial.rx(1)
if prompt == '$':
    #
    #  Get all registers content
    #
    for r in nrf_registers:
        rn, stx, nrx = r
        ntx = len(stx)
        ra = ord(stx[0])
        cout("Register %-10s: 0x%02X = " % (rn,ra))
        flushout()
        serial.tx('='+chr(ntx)+chr(nrx)+stx)
        r = serial.rx(nrx+1)
        if r:
            if r.endswith(prompt):
                cout( s2hex(r[:-1])+'\n' )
            else:
                die(" received `%s`\n" % s2hex(r))
        else:
            cout("no response.\n")
else:
    cout("Unexpected prompt: %s\n" % s2hex(prompt))

serial.close()

File: atmel/avr/examples/09-3-uart-spi-nrf24l01+/main.c

/*  This file is part of the HWA project.
 *  Copyright (c) 2012,2015,2017 Christophe Duparquet.
 *  All rights reserved. Read LICENSE.TXT for details.
 */

/**
 * @example
 * 
 * @par A simple interface between UART and SPI master connected to a nRF24L01
 * 
 * @par nRF24L01+ module wiring
 *
 *                   Gnd  [1](2)  Vcc
 *            Gnd <- CE   (3)(4)  CSN  -> SS
 *            SCL <- SCK  (5)(6)  MOSI -> MOSI
 *           MISO <- MISO (7)(8)  IRQ
 *
 * @par Test application
 *
 *     ./main.py
 * 
 * should display:
 *
 *     Register CONFIG    : 0x00 = 08
 *     Register EN_AA     : 0x01 = 3F
 *     Register EN_RX_ADDR: 0x02 = 03
 *     Register SETUP_AW  : 0x03 = 03
 *     Register SETUP_RETR: 0x04 = 03
 *     Register RF_CH     : 0x05 = 02
 *     Register RF_SETUP  : 0x06 = 0F
 *     Register STATUS    : 0x07 = 0E
 *     Register RX_ADDR_P0: 0x0A = E7 E7 E7 E7 E7
 *     Register RX_ADDR_P1: 0x0B = C2 C2 C2 C2 C2
 *     Register DYNPD     : 0x1C = 00
 *     Register FEATURE   : 0x1D = 00
 * 
 * * `BOARD_H` is the name of the target board header file. It can be defined at
 * compile time via the command line. For example, `make BOARD=nanodccduino`
 * will define `BOARD_H` as `<boards/nanodccduino.h>`. See @ref atmelavr_boards for the
 * board definitions provided with HWA.
 * 
 * @par main.c
 **/

#include BOARD_H

#if defined ARDUINO
/*
 *  If a pin other than PIN_SS is used, and PIN_SS is an input, it must be held
 *  high otherwise the SPI will be automatically turned into slave mode by
 *  hardware as soon as PIN_SS goes low.
 */
#  define NRF_CSN	PIN_SS
#else
#  error That should not happen
#endif

#define UART		uart0


/*  Transfer 1 byte on the SPI.
 *  This is used for reading as well as for writing.
 *  Return the received byte.
 */
uint8_t spi ( uint8_t v )
{
  hw( write, spi0, v );
  while( !hw( read, (spi0,irq) ) ) {}
  return hw( read, spi0 );
}


/*  Wait for one byte received on the UART and read it.
 */
uint8_t uart_rx()
{
  while ( !hw( read, (UART,irq,rxc) ) ) {}
  return hw( read, UART );
}


/*  Wait until the transmit buffer of the UART is not full, then write one byte
 *  into it.
 */
void uart_tx( uint8_t c )
{
  while ( !hw( read, (UART,irq,txqnf) ) ) {}
  hw( write, UART, c );
}


int
main ( )
{
  /*  Create a HWA context to collect the hardware configuration
   *  Preload this context with RESET values
   */
  hwa( begin, reset );

  /*  Configure the software UART
   */
  hwa( configure, UART,
       bps,	  115200 );

  /*  Configure the SPI as master
   */
  hwa( configure,     spi0,
       function,      master,
       clock,	      ioclk / 128,
       /* mode,          0, */
       idle_state,    low,
       sampling_edge, rising,
       data_order,    msb_first );

  /*  Configure nRF CSN pin
   */
  hwa( configure, NRF_CSN, mode, digital_output );
  hwa( write, NRF_CSN, 1 );

  /*  Write this configuration into the hardware
   */
  hwa( commit );

  /*  Process commands from host
   */
  for(;;) {

    /*	Prompt
     */
    uart_tx( '$' );

    /*	The host sends commands starting with '=' and followed by:
     *	  * the number of bytes to send to SPI slave (1 byte)
     *	  * the number of bytes to read (1 byte)
     *	  * the bytes to send
     */
    uint8_t c = uart_rx();
    if ( c == '=' ) {

      /*  Number of bytes to send to SPI slave
       */
      uint8_t ntx = uart_rx();
      if ( ntx < 1 || ntx > 33 )
	goto error ;

      /*  Number of bytes to send back to talker
       */
      uint8_t nrx = uart_rx();
      if ( nrx > 32 )
	goto error ;

      /*  Select SPI slave and send data
       */
      hw( write, NRF_CSN, 0 );
      while ( ntx-- ) {
	c = uart_rx();
	spi( c );
      }

      /*  Send reply to host and deselect SPI slave
       */
      while ( nrx-- ) {
	c = spi(0);
	uart_tx(c);
      }
      hw( write, NRF_CSN, 1 );
    }
    else {
      /*
       *  First byte of command must be '='. Send '!' as error indicator and
       *  wait for '\n' as error acknowledgement.
       */
      do {
      error:
	uart_tx('!');
	c = uart_rx();
      } while ( c != '\n' ) ;
    }
  }
}
